atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/store/UnreliableTimestampManagerTest.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.workload.store;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.google.common.collect.Iterables;
+import com.palantir.lock.client.TimestampManager;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.v2.TimelockService;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.Named;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public class UnreliableTimestampManagerTest {
+
+    @Mock
+    private TimelockService timelockService;
+
+    @Test
+    public void randomlyIncreaseTimestampRandomlyIncreasesTimestamp() {
+        List<Long> allValues = new ArrayList<>(List.of(0L));
+        UnreliableTimestampManager timestampManager =
+                UnreliableTimestampManager.create(timelockService, () -> Iterables.getLast(allValues), allValues::add);
+        for (int i = 0; i < 10; i++) {
+            timestampManager.randomlyIncreaseTimestamp();
+        }
+
+        assertListRandomlyIncreasesInValue(allValues);
+    }
+
+    @Test
+    public void randomlyIncreaseTimestampRandomlyIncreasesTimestampWithMultipleThreads() throws InterruptedException {
+        ArrayList<Long> allValues = new ArrayList<>(List.of(0L));
+        UnreliableTimestampManager timestampManager = createThreadsafeArrayBackedTimestampManager(allValues);
+
+        ExecutorService executor = Executors.newCachedThreadPool();
+        try {
+            for (int i = 0; i < 10; i++) {
+                executor.execute(timestampManager::randomlyIncreaseTimestamp);
```

Comment:
I typically prefer using ListenableFutures, combining them into and awaiting that at the end of whatever you want. That way you can call shutdownNow rightaway (and if it returns a non-empty list and you want to be extra careful you can assert that's empty?)